### ISSUE TITLE
From which textbook/article is the transfer entropy obtained from?

### ISSUE THREAD

**Issue by Datseris** [JuliaDynamics/CausalityTools.jl#119]
Hi there,

to learn about transfer entropy I've studied the original publication by Schreiber. I've noticed that the docs have entirely different notation, while using some of the same symbols (e.g. $k$ is used in both papers with different meaning). 

What is the review/textbook/paper that your notation is coming from? I'd like to read that to avoid further confusion while reading the docs.

**Comment by kahaaga** [comment]
There are different notations for the transfer entropy from different papers. We will have to decide on one here, and refer readers to a publication that uses that notation. There are three marginals (four, if conditioning is included) to consider for transfer entropy: 

- the future of the target variable. 
- the present/past of the target variable
- the present/past of the source variable
- the present/past of the conditional variable

## Suggestion 

Personally, I am in favor of using the following notation: `T` for the target future, `Tpp` for the target present/past, `Spp` for the source present/past and `Cpp` for conditional present/past. This way, the user doesn't have to always look up which symbol corresponds to which marginal. 

In the code, one would then use `dTpp`/`τTpp` to indicate the dimension/embedding lag of the delay embedding for the `Tpp` marginal, and the same for the remaining marginals.

I will have a discussion with the guys in the group on what to go for.  Input here is also very welcome! 

(note to self: a decision on notation must also be made before I can finish up #121).

**Comment by Datseris** [comment]
I find these subscipts not so helpful, because they introduce a lot of clutter in the expressions, and all could be avoided by renaming `T` to `F` (for future) or something along these lines. If someone has studied the definition of the transfer entropy, it is clear that the future means the future of the "target", and it is not absolutely necessary to use the same symbol. Target and source are useful and I agree with them, and thus I would suggest the consistend `F, T, S, C` throughout. 

The difficulty for me is that each one of these four is also accompanied with 2 parameters (making 8 in total): the embedding dimension and the delay time. How should we choose these in a simple manner? we could use the same symbol for all of them, follwed by the name: `dF, τF`, like you suggest. I think this will make the interface clear and easier to learn.

**Comment by Datseris** [comment]
Haha! Julia allows you to use unicode. I think it makes sense to just use script `\mathcal{T}` and in code `𝒯` (which is `\scrT`)

**Comment by kahaaga** [comment]
Using `𝒯` for the target future is a good idea! So:

- `T`: target present/past. In the code, `dT` denotes the dimension and `τT` the embedding lag.
- `S`: source present/past. In the code, `dS` denotes the dimension and `τS` the embedding lag.
- `C`: conditional present/past. In the code, `dC` denotes the dimension and `τC` the embedding lag.
- `𝒯`: target future. In the code, `d𝒯` denotes the dimension and `η𝒯` the prediction lag. Use the distinct symbol `η` for the lag here, because it is the prediction lag, not the regular embedding lag (see explanation below). Using `η` for the prediction lag also aligns with the notation for our predictive asymmetry method in our upcoming paper.

Are we on the same page?

_Note: Most often, for transfer entropy, `T`, `S` and `C` marginals always include the instantaneous lag. Therefore, `dT` must be `>= 2` for the embedding delay `τT` to have any effect. The same applies to the `S` and `C` marginals. For the `𝒯` marginal, the instantaneous lag is not included, so `η` is always `>= 1`._

**Comment by Datseris** [comment]
Yeap I think we are, although I am confused a bit with `η`. Can you explain it in a bit more detail? In the original paper by schereiber he used `τ=1` de facto for both source and target, while he used the symbols `k` for the dimension of the target and `l` for the dimension of the source. For him the future always had a time lag of "+1", meaning that you consider how the exactly next value depends on the `k` previous values of the target and the `l` previous values of the source.

What is my understanding is that one can (of course) use any `τ` greater than 1 and that SChreiber just didn't mention this in his paper. But what does `η` do? Also, in the original paper `d𝒯` is by definition 0, as there is no embedding happening for the future. I am guessing that you generalize this to arbitrary forwards embedding for the future of the target, correct?

**Comment by kahaaga** [comment]
Correct, in this implementation, any marginal may be scalar-valued vectors (i.e. the raw time series, but just shifted) or arbitrary embeddings of those time series. This applies both to the `T`, `S`, `C` and `𝒯` (future) marginals. The difference in my implementation is thatfor the future of the target (`𝒯`), the embedding is done forwards-in-time, not backwards-in-time (as for `T`, `S` and `C`). 

Think `𝒯 = {(x(t+η), x(t + 2*η), ...)}`, vs. `T = {(x(t - τ), x(t - 2*τ), ...)}` (opposite signs). If `d𝒯 = 1`, then `𝒯 = {(x(t+η)}` (not including instantaneous lag), and if `dT = 1`, then `T = {(x(t)}` (including instantaneous lag).

In principle one can use any `τ`, and it doesn't need not be fixed. Some authors even optimise `τ` individually for each component of each marginal. That is, `T = {(x(t - τ_1), x(t - τ_2), ...)}` with `τ_1, τ_2, ...` optimised individually according to some criterion.

Computing transfer entropy for many prediction lags `η`, while keeping the embedding delays fixed, is often (I would say always) required for real world data. The intrinsic time scale of an observed system and the time-delay of the causal influence is unknown a priori, so it may not be sufficient to consider only how including information about the `source` variable affects our predictions about the exactly next value of `target`. To capture the causal influence, one may also need to quantify how inclusion of past/present values of `source` affects the `1:η_max` next values of `target` (there is not a 1:1 correspondence between the causal lag and the prediction lag `η`, so multiple `η`s should be tested). 

For example, [Krakovska et al. (2018)](https://journals.aps.org/pre/abstract/10.1103/PhysRevE.97.042207) computes transfer entropy over multiple prediction lags (`η`) with fixed `τT` and `τT` and considers the `η`-averaged transfer entropy.

**Comment by Datseris** [comment]
Thanks, this information is extremely useful for me! Especially the practical application tips you give! 

There is only one point that I consider not perfectly intuitive is:
```
If d𝒯 = 1, then 𝒯 = {(x(t+η)} (not including instantaneous lag), and if dT = 1, then T = {(x(t)} (including instantaneous lag).
```
I now understand why it happens (because it is mandatory to have some future value in `𝒯` to calculate transfer entropy), its just something one has to thing for a couple of minutes.

But so far all great! This notation is getting clearer and clearer!

**Comment by kahaaga** [comment]
Good to hear! It is hard to phrase these things in an intuitive manner. I think the notation above will help, though. We just need to incorporate clear explanations in the package documentation too, so that people don't have to search in old issues for it. 

In practice, for synthetic systems with instantaneous effects or for real data with poor temporal resolution, there might actually be instantaneous information transfer, but this is not something that TE can pick up and some other method show be used. For transfer entropy you need `η > 1`to avoid overlap between the `𝒯` and `T` marginals.

**Comment by Datseris** [comment]
It might be interesting for you to see the way I've decided to write these things in the book I am currently co-authoring with U. Parlitz about nonlinear dynamics and chaos. We decided to have 2-4 pages in mutual info / transfer entropy / causality some time ago, and after our (extremely helpful, I am citing you in the acknowledgements) discussions this is the final product I decided to use:


![image](https://user-images.githubusercontent.com/19669089/73358615-52867c00-429f-11ea-875f-74d1038d999d.png)

It is **much** more verbose that then version that we discussed here, but I thought it makes sense to explicitly include the embedding parameters `d, τ` in the expression (I actually use `γ` similarly to how I use it in DynamicalSystems.jl).

**Comment by Datseris** [comment]
I've tagged you at the `genembed` new style that I am using in DelayEmbeddings.jl. I also use this style in this book I've mentioned above, and thus what I've pasted is completely obsolete. 

Once the generalized embedding is introduced, it is very straight forward and natural to say something like 

![image](https://user-images.githubusercontent.com/19669089/80963117-a066cf00-8e0e-11ea-913d-7435b6a8f073.png)

and it is also trivial to extend this for a conditional timeseries `z`.
